<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Sequence Matrix by gaurav</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Sequence Matrix</h1>
      <h2 class="project-tagline">Assemble multigene/multitaxon phylogenetic datasets easily</h2>
      <a href="https://github.com/gaurav/taxondna" class="btn">View on GitHub</a>
      <a href="https://github.com/gaurav/taxondna/releases/latest" class="btn">Download latest release</a>
      <a href="https://github.com/gaurav/taxondna/zipball/master" class="btn">Download source as .zip</a>
      <a href="https://github.com/gaurav/taxondna/tarball/master" class="btn">Download source as .tar.gz</a>
    </section>

    <section class="main-content">

      <p>Sequence Matrix allows you to easily concatenate gene datasets. A spreadsheet-like interface displays what you've assembled so far; TNT, Nexus, FASTA and Mega files can be dragged into the application to add them. Your entire dataset can be exported as TNT or Nexus files.</p>

<p>To find out more, please see our <a href="http://www.youtube.com/watch?v=Adwr-CZdNOM">introductory screencast</a> of Sequence Matrix on YouTube. We've also <a href="http://www.slideshare.net/mrvaidya/sequence-matrix-gene-concatenation-made-easy-4797559">made available a presentation on Sequence Matrix</a> (delivered to the Hennig XXVIII conference in June 2009) and <a href="http://dx.doi.org/10.1111/j.1096-0031.2010.00329.x">the paper we wrote</a> describing Sequence Matrix in Cladistics. </p>

<center>
    <iframe width="560" height="315" src="https://www.youtube.com/embed/Adwr-CZdNOM" frameborder="0" allowfullscreen></iframe>
</center>

<h1>
<a id="for-users-of-macos-x" class="anchor" href="#for-users-of-macos-x" aria-hidden="true"><span class="octicon octicon-link"></span></a>For users of MacOS X
</h1>

<p>If you get an "unidentified developer" error, please right-click on the file and select "Open": you should get the some error, but with an "Open" button that allows you to bypass it. Alternatively, you can change your security settings: go to System Preferences on your Mac, click on "Security &amp; Privacy" and go to the "General" tab. If you set "All apps downloaded from:" to "Anywhere", Sequence Matrix should run without any problem.</p>

<h1>
<a id="testimonials" class="anchor" href="#testimonials" aria-hidden="true"><span class="octicon octicon-link"></span></a>Testimonials</h1>

<p>Here are some reviews of Sequence Matrix we've spotted around the internet.</p>

<ul>
<li><p>"Thanks so much. This software has saved me many﻿ hours of tedious work." -- <a href="http://www.youtube.com/user/evanmelstad">evanmelstad</a> on <a href="http://www.youtube.com/comment?lc=bOaM8sE6UvMNa2ZilLgH45-Uti1e3fwDMQ2dXEtzNXQ">YouTube</a>.</p></li>
<li>
<p>"It is a real pain combining matrices from different loci, particularly when you have taxa represented in one matrix and not in another. I’ve usually done this by hand and it gets tedious. I recently stumbled across a Java tool that makes this process quite simple: SequenceMatrix." -- <a href="http://lacertilia.com/?p=1420">The Glor Lab</a>, University of Rochester. </p>
</li>

<li>
<p>"... the 12 individual gene alignments were concatenated using the software SequenceMatrix v1.7.6 (Vaidya et al., 2011) to create a master alignment of 10,836 bp total, equivalent to roughly 65% of the mtDNA genome." -- <em><a href="http://dx.doi.org/10.1016/j.ympev.2011.03.025">Pleistocene diversification of living squirrel monkeys (Saimiri spp.) inferred from complete mitochondrial genome sequences</a></em> by Kenneth L. Chioua, Luca Pozzia, Jessica W. Lynch Alfaroc and Anthony Di Fiore. </p>
</li>
</ul>

<h1>What about Species Identifier?</h1>

<p>Species Identifier is no longer being actively maintained. If you want to
repeat the analyses in 
<a href="http://dx.doi.org/10.1080/10635150600969864">Meier <em>et al</em>, 2006</a>,
you can download <a href="https://github.com/gaurav/taxondna/releases/latest">the latest release</a>.
</p>

<h1>
<a id="problems-let-us-know" class="anchor" href="#problems-let-us-know" aria-hidden="true"><span class="octicon octicon-link"></span></a>Problems? Let us know!</h1>

<p>If you find any bugs in this or any other release, please <a href="https://github.com/gaurav/taxondna/issues">report them to us</a>! The source code for this project is available <a href="http://github.com/gaurav/taxondna">on GitHub</a>.</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/gaurav/taxondna">Sequence Matrix</a> is maintained by <a href="https://github.com/gaurav">gaurav</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-17621072-5");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>

